File: cpp/medium/FindKthLargest.cpp
#include <climits>
#include <vector>
using namespace std;

int quick(vector<int> &nums, int left, int right, int k) {
  int next_left = left;
  int next_right = right;
  int cur = nums[left];
  while (left < right) {
    while (left < right && nums[right] > cur) {
      right--;
    }
    if (left < right) {
      nums[left] = nums[right];
      left++;
    }
    while (left < right && nums[left] < cur) {
      left++;
    }
    if (left < right) {
      nums[right] = nums[left];
      right--;
    }
  }
  nums[left] = cur;
  if (left == k)
    return cur;
  if (left < k)
    return quick(nums, left + 1, next_right, k);
  return quick(nums, next_left, left - 1, k);
}

int findKthLargest(vector<int> &nums, int k) {
  return quick(nums, 0, nums.size() - 1, nums.size() - k);
}

File: go/middle/findKthLargest.go
package main

func findKthLargest(nums []int, k int) int {
	rs := 0
	var quickSort func(left, right int)
	quickSort = func(left, right int) {
		if left > right {
			return
		}
		newLeft := left
		newRight := right
		split := nums[left]
		for left < right {
			for left < right && nums[right] > split {
				right--
			}
			if left < right {
				nums[left] = nums[right]
				left++
			}
			for left < right && nums[left] < split {
				left++
			}
			if left < right {
				nums[right] = nums[left]
				right--
			}
		}
		nums[left] = split
		if left == len(nums)-k {
			rs = nums[left]
			return
		} else if left < len(nums)-k {
			quickSort(left+1, newRight)
		} else {
			quickSort(newLeft, left-1)
		}
	}
	quickSort(0, len(nums)-1)
	return rs
}
